arangod/RestHandler/RestDocumentHandler.cpp

```diff
@@ -269,67 +267,68 @@ bool RestDocumentHandler::readSingleDocument(bool generateBody) {
   VPackSlice search = builder.slice();
 
   // find and load collection given by name or identifier
-  auto trx = createTransaction(collection, AccessMode::Type::READ);
+  _activeTrx = createTransaction(collection, AccessMode::Type::READ);
 
-  trx->addHint(transaction::Hints::Hint::SINGLE_OPERATION);
+  _activeTrx->addHint(transaction::Hints::Hint::SINGLE_OPERATION);
 
   // ...........................................................................
   // inside read transaction
   // ...........................................................................
 
-  Result res = trx->begin();
+  Result res = _activeTrx->begin();
 
   if (!res.ok()) {
     generateTransactionError(collection, res, "");
-    return false;
+    return RestStatus::DONE;
   }
 
-  OperationResult result = trx->document(collection, search, options);
-
-  res = trx->finish(result.result);
-
-  if (!result.ok()) {
-    if (result.is(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND)) {
-      generateDocumentNotFound(collection, key);
-      return false;
-    } else if (ifRid != 0 && result.is(TRI_ERROR_ARANGO_CONFLICT)) {
-      generatePreconditionFailed(result.slice());
-    } else {
-      generateTransactionError(collection, res, key);
-    }
-    return false;
-  }
+  return waitForFuture(
+      _activeTrx->documentAsync(collection, search, options).thenValue([=](OperationResult opRes) {
+        auto res = _activeTrx->finish(opRes.result);
+
+        if (!opRes.ok()) {
+          if (opRes.is(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND)) {
+            generateDocumentNotFound(collection, key);
+            return false;
```

Comment:
micro optimization: this line could be removed. even without this line, the next instruction will be `return false`.